{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1de7541c-2156-4e5f-bad9-418fc6791d88",
   "metadata": {},
   "outputs": [],
   "source": [
    "include <- function(pkg) {\n",
    "  if (!suppressMessages(require(pkg, character.only = TRUE)))\n",
    "    install.packages(pkg, character.only = TRUE)\n",
    "  suppressMessages(library(pkg, pkg, character.only = TRUE))\n",
    "}\n",
    "include(\"ShortRead\")\n",
    "include(\"plyr\")\n",
    "include(\"magick\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5e7bdb83-0acc-4b7c-8078-1a25800c9db7",
   "metadata": {},
   "outputs": [],
   "source": [
    "BASE_PATH_NR1 = \"/data/samples/AIRR-Seq/OURS/S3987Nr1\"\n",
    "BASE_PATH_NR2 = \"/data/samples/AIRR-Seq/OURS/S3987Nr2\"\n",
    "\n",
    "SAMPLES = list(\n",
    "    list(base_path=BASE_PATH_NR1, name=\"S3987Nr1-PBMC1_heavy\"),\n",
    "    list(base_path=BASE_PATH_NR1, name=\"S3987Nr1-PBMC1_light\"),\n",
    "#     list(base_path=BASE_PATH_NR1, name=\"S3987Nr1-RAMOS_heavy\"),\n",
    "    list(base_path=BASE_PATH_NR1, name=\"S3987Nr1-RAMOS_light\")\n",
    "#     list(base_path=BASE_PATH_NR2, name=\"S3987Nr2-PBMC1_heavy\"),\n",
    "#     list(base_path=BASE_PATH_NR2, name=\"S3987Nr2-PBMC1_light\"),\n",
    "#     list(base_path=BASE_PATH_NR2, name=\"S3987Nr2-RAMOS_heavy\"),\n",
    "#     list(base_path=BASE_PATH_NR2, name=\"S3987Nr2-RAMOS_light\")\n",
    ")\n",
    "\n",
    "EXCLUDE_FROM_DIVERSITY_ANALYSIS = c(\n",
    "    \"S3987Nr1-RAMOS_heavy\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "35dd51fc-a9ea-4e20-9163-437850b4b0a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Helpers\n",
    "get_path_sample <- function(sample_name, base_path = BASE_PATH) {\n",
    "    return(paste0(base_path, \"/\", sample_name))\n",
    "}\n",
    "\n",
    "get_path_fastq <- function(sample_name, type = \"raw\", base_path = BASE_PATH) {\n",
    "    \n",
    "    if (type == \"raw\") {\n",
    "        return(\n",
    "            paste0(\n",
    "                get_path_sample(sample_name, base_path),\n",
    "                \"/\", sample_name, \".1.fastq\"\n",
    "            )\n",
    "        )\n",
    "    } else if (type == \"primer_pass\") {\n",
    "        suffix = \".1_primers-pass.fastq\"\n",
    "    } else if (type == \"pair_pass\"){\n",
    "        suffix = \".1_primers-pass_pair-pass.fastq\"\n",
    "    } else if (type == \"under2\") {\n",
    "        suffix = \"_under-2.fastq\"\n",
    "    } else if (type == \"atleast2\") {\n",
    "        suffix = \"_atleast-2.fastq\"\n",
    "    } else {\n",
    "        stop(paste0(\"Unsupported fastq file type: \", type))\n",
    "    }\n",
    "    \n",
    "    return(\n",
    "        paste0(\n",
    "            get_path_sample(sample_name, base_path),\n",
    "            \"/\", \"presto\",\n",
    "            \"/\", sample_name, suffix\n",
    "        )\n",
    "    )\n",
    "}\n",
    "\n",
    "get_path_igblast <- function(sample_name, base_path = BASE_PATH) {\n",
    "    return(\n",
    "        paste0(\n",
    "            get_path_sample(sample_name, base_path), \n",
    "            \"/\", \"changeo_igblast\", \n",
    "            \"/\", sample_name, \"_db-pass_with_translation.tsv\"\n",
    "        )\n",
    "    )\n",
    "}\n",
    "\n",
    "get_path_clones <- function(sample_name, clone_file = \"groups\", base_path = BASE_PATH) {\n",
    "    \n",
    "    if (clone_file == \"groups\") {\n",
    "        suffix = \"_vjl_groups.tsv\"\n",
    "    } else if (clone_file == \"clones\") {\n",
    "        suffix = \"_with_clones.tsv\"\n",
    "    } else if (clone_file == \"counts\") {\n",
    "        return(\n",
    "            paste0(\n",
    "                get_path_sample(sample_name, base_path),\n",
    "                \"/\", \"clones\", \n",
    "                \"/\", \"diversity\",\n",
    "                \"/\", sample_name, \"_clone_counts.tsv\"\n",
    "            )\n",
    "        )\n",
    "    } else {\n",
    "        stop(paste0(\"Unknown clone_file type: \", clone_file))\n",
    "    }\n",
    "    \n",
    "    return(\n",
    "        paste0(\n",
    "            get_path_sample(sample_name, base_path),\n",
    "            \"/\", \"clones\", \n",
    "            \"/\", sample_name, suffix\n",
    "        )\n",
    "    )\n",
    "}\n",
    "\n",
    "get_path_coverages <- function(sample_name, base_path = BASE_PATH) {\n",
    "    return(\n",
    "        paste0(\n",
    "            get_path_sample(sample_name, base_path),\n",
    "            \"/\", \"clones\",\n",
    "            \"/\", \"diversity\",\n",
    "            \"/\", sample_name, \"_coverages.tsv\"\n",
    "        )\n",
    "    )\n",
    "}\n",
    "\n",
    "get_path_diversity <- function(sample_name, base_path = BASE_PATH) {\n",
    "    return(\n",
    "        paste0(\n",
    "            get_path_sample(sample_name, base_path),\n",
    "            \"/\", \"clones\",\n",
    "            \"/\", \"diversity\",\n",
    "            \"/\", sample_name, \"_diversity.tsv\"\n",
    "        )\n",
    "    )\n",
    "}\n",
    "\n",
    "get_path_pngs <- function(sample_name, img_type = \"abundancy\", base_path = BASE_PATH) {\n",
    "    \n",
    "    if (img_type == \"abundancy\") {\n",
    "        suffix = \"_abundancy_curve.png\"\n",
    "    } else if (img_type == \"diversity\") {\n",
    "        suffix = \"_diversity.png\"\n",
    "    } else if (img_type == \"clones\") {\n",
    "        return(\n",
    "            paste0(\n",
    "                get_path_sample(sample_name, base_path),\n",
    "                \"/\", \"clones\",\n",
    "                \"/\", sample_name, \"_with_clones.png\"\n",
    "            )\n",
    "        )\n",
    "    } else {\n",
    "        stop(\"Image type = \", img_type, \" is not supported.\")\n",
    "    }\n",
    "    \n",
    "    return(\n",
    "        paste0(\n",
    "            get_path_sample(sample_name, base_path),\n",
    "            \"/\", \"clones\",\n",
    "            \"/\", \"diversity\",\n",
    "            \"/\", sample_name, suffix\n",
    "        )\n",
    "    )\n",
    "}\n",
    "\n",
    "read_tsv <- function(filepath) {\n",
    "    return(read.csv(filepath, sep='\\t'))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47b738dd-92d8-4ef9-b452-bccad31e7663",
   "metadata": {},
   "source": [
    "### Characterize sequencing results for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7540eb42-8db2-4864-a627-55d0865f7e88",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"/data/samples/AIRR-Seq/OURS/S3987Nr1/S3987Nr1-PBMC1_heavy/presto/S3987Nr1-PBMC1_heavy_atleast-2.fastq\"\n"
     ]
    }
   ],
   "source": [
    "# fp = get_path_fastq(SAMPLES[[1]]$name, type = \"atleast2\", SAMPLES[[1]]$base_path)\n",
    "# print(fp)\n",
    "# qa_summary = ShortRead::qa(fp, type = \"fastq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "451743c5-63df-4c6b-8da1-d883b7e8b542",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: FastqQA(10)\n",
       "QA elements (access with qa[[\"elt\"]]):\n",
       "  readCounts: data.frame(1 3)\n",
       "  baseCalls: data.frame(1 5)\n",
       "  readQualityScore: data.frame(512 4)\n",
       "  baseQuality: data.frame(95 3)\n",
       "  alignQuality: data.frame(1 3)\n",
       "  frequentSequences: data.frame(50 4)\n",
       "  sequenceDistribution: data.frame(22 4)\n",
       "  perCycle: list(2)\n",
       "    baseCall: data.frame(1629 4)\n",
       "    quality: data.frame(5857 5)\n",
       "  perTile: list(2)\n",
       "    readCounts: data.frame(0 4)\n",
       "    medianReadQualityScore: data.frame(0 4)\n",
       "  adapterContamination: data.frame(1 1)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# report_html(qa_summary, \"test.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bf98f821-9610-4623-a6fd-257505f6e4c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_reads_fastq <- function(fastq_path) {\n",
    "    return(\n",
    "        ShortRead::countFastq(fastq_path)$records\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "021306a9-4829-4bb3-9696-085e97c5f035",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 x 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>sample_name</th><th scope=col>raw</th><th scope=col>primer_pass</th><th scope=col>pair_pass</th><th scope=col>under2</th><th scope=col>atleast2</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>S3987Nr1-PBMC1_heavy</td><td>293087 </td><td>293087 </td><td>293087 </td><td>127350 </td><td>14008 </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>S3987Nr1-PBMC1_light</td><td>264459 </td><td>264459 </td><td>264459 </td><td>108203 </td><td>13384 </td></tr>\n",
       "\t<tr><th scope=row>3</th><td>S3987Nr1-RAMOS_heavy</td><td>160952 </td><td>160952 </td><td>160952 </td><td>28     </td><td>5     </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>S3987Nr1-RAMOS_light</td><td>69991  </td><td>69991  </td><td>69991  </td><td>25974  </td><td>3791  </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>S3987Nr2-PBMC1_heavy</td><td>4130140</td><td>4130140</td><td>4130140</td><td>1336389</td><td>187452</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>S3987Nr2-PBMC1_light</td><td>3756935</td><td>3756935</td><td>3756935</td><td>1099571</td><td>170346</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 x 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & sample\\_name & raw & primer\\_pass & pair\\_pass & under2 & atleast2\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & S3987Nr1-PBMC1\\_heavy & 293087  & 293087  & 293087  & 127350  & 14008 \\\\\n",
       "\t2 & S3987Nr1-PBMC1\\_light & 264459  & 264459  & 264459  & 108203  & 13384 \\\\\n",
       "\t3 & S3987Nr1-RAMOS\\_heavy & 160952  & 160952  & 160952  & 28      & 5     \\\\\n",
       "\t4 & S3987Nr1-RAMOS\\_light & 69991   & 69991   & 69991   & 25974   & 3791  \\\\\n",
       "\t5 & S3987Nr2-PBMC1\\_heavy & 4130140 & 4130140 & 4130140 & 1336389 & 187452\\\\\n",
       "\t6 & S3987Nr2-PBMC1\\_light & 3756935 & 3756935 & 3756935 & 1099571 & 170346\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 x 6\n",
       "\n",
       "| <!--/--> | sample_name &lt;chr&gt; | raw &lt;chr&gt; | primer_pass &lt;chr&gt; | pair_pass &lt;chr&gt; | under2 &lt;chr&gt; | atleast2 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | S3987Nr1-PBMC1_heavy | 293087  | 293087  | 293087  | 127350  | 14008  |\n",
       "| 2 | S3987Nr1-PBMC1_light | 264459  | 264459  | 264459  | 108203  | 13384  |\n",
       "| 3 | S3987Nr1-RAMOS_heavy | 160952  | 160952  | 160952  | 28      | 5      |\n",
       "| 4 | S3987Nr1-RAMOS_light | 69991   | 69991   | 69991   | 25974   | 3791   |\n",
       "| 5 | S3987Nr2-PBMC1_heavy | 4130140 | 4130140 | 4130140 | 1336389 | 187452 |\n",
       "| 6 | S3987Nr2-PBMC1_light | 3756935 | 3756935 | 3756935 | 1099571 | 170346 |\n",
       "\n"
      ],
      "text/plain": [
       "  sample_name          raw     primer_pass pair_pass under2  atleast2\n",
       "1 S3987Nr1-PBMC1_heavy 293087  293087      293087    127350  14008   \n",
       "2 S3987Nr1-PBMC1_light 264459  264459      264459    108203  13384   \n",
       "3 S3987Nr1-RAMOS_heavy 160952  160952      160952    28      5       \n",
       "4 S3987Nr1-RAMOS_light 69991   69991       69991     25974   3791    \n",
       "5 S3987Nr2-PBMC1_heavy 4130140 4130140     4130140   1336389 187452  \n",
       "6 S3987Nr2-PBMC1_light 3756935 3756935     3756935   1099571 170346  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "res <- sapply(\n",
    "    SAMPLES,\n",
    "    function(sample) {\n",
    "        counts = cbind(\n",
    "            sample$name,\n",
    "            count_reads_fastq(get_path_fastq(sample$name, type = \"raw\", sample$base_path)),\n",
    "            count_reads_fastq(get_path_fastq(sample$name, type = \"primer_pass\", sample$base_path)),\n",
    "            count_reads_fastq(get_path_fastq(sample$name, type = \"pair_pass\", sample$base_path)),\n",
    "            count_reads_fastq(get_path_fastq(sample$name, type = \"under2\", sample$base_path)),\n",
    "            count_reads_fastq(get_path_fastq(sample$name, type = \"atleast2\", sample$base_path))\n",
    "        )\n",
    "        return(counts)\n",
    "    }\n",
    ")\n",
    "\n",
    "res <- data.frame(t(res))\n",
    "colnames(res) = c(\"sample_name\", \"raw\", \"primer_pass\", \"pair_pass\", \"under2\", \"atleast2\")\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cb1a593f-b999-4808-9a21-ffe89d029b74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 7 x 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>sample_name</th><th scope=col>raw</th><th scope=col>primer_pass</th><th scope=col>pair_pass</th><th scope=col>under2</th><th scope=col>atleast2</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td>293087 </td><td>293087 </td><td>293087 </td><td>127350 </td><td>14008 </td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td>264459 </td><td>264459 </td><td>264459 </td><td>108203 </td><td>13384 </td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_heavy</td><td>160952 </td><td>160952 </td><td>160952 </td><td>28     </td><td>5     </td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_light</td><td>69991  </td><td>69991  </td><td>69991  </td><td>25974  </td><td>3791  </td></tr>\n",
       "\t<tr><td>S3987Nr2-PBMC1_heavy</td><td>4130140</td><td>4130140</td><td>4130140</td><td>1336389</td><td>187452</td></tr>\n",
       "\t<tr><td>S3987Nr2-PBMC1_light</td><td>3756935</td><td>3756935</td><td>3756935</td><td>1099571</td><td>170346</td></tr>\n",
       "\t<tr><td>S3987Nr2-RAMOS_heavy</td><td>2273707</td><td>2273707</td><td>2273707</td><td>511006 </td><td>82177 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 7 x 6\n",
       "\\begin{tabular}{llllll}\n",
       " sample\\_name & raw & primer\\_pass & pair\\_pass & under2 & atleast2\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t S3987Nr1-PBMC1\\_heavy & 293087  & 293087  & 293087  & 127350  & 14008 \\\\\n",
       "\t S3987Nr1-PBMC1\\_light & 264459  & 264459  & 264459  & 108203  & 13384 \\\\\n",
       "\t S3987Nr1-RAMOS\\_heavy & 160952  & 160952  & 160952  & 28      & 5     \\\\\n",
       "\t S3987Nr1-RAMOS\\_light & 69991   & 69991   & 69991   & 25974   & 3791  \\\\\n",
       "\t S3987Nr2-PBMC1\\_heavy & 4130140 & 4130140 & 4130140 & 1336389 & 187452\\\\\n",
       "\t S3987Nr2-PBMC1\\_light & 3756935 & 3756935 & 3756935 & 1099571 & 170346\\\\\n",
       "\t S3987Nr2-RAMOS\\_heavy & 2273707 & 2273707 & 2273707 & 511006  & 82177 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 7 x 6\n",
       "\n",
       "| sample_name &lt;chr&gt; | raw &lt;chr&gt; | primer_pass &lt;chr&gt; | pair_pass &lt;chr&gt; | under2 &lt;chr&gt; | atleast2 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| S3987Nr1-PBMC1_heavy | 293087  | 293087  | 293087  | 127350  | 14008  |\n",
       "| S3987Nr1-PBMC1_light | 264459  | 264459  | 264459  | 108203  | 13384  |\n",
       "| S3987Nr1-RAMOS_heavy | 160952  | 160952  | 160952  | 28      | 5      |\n",
       "| S3987Nr1-RAMOS_light | 69991   | 69991   | 69991   | 25974   | 3791   |\n",
       "| S3987Nr2-PBMC1_heavy | 4130140 | 4130140 | 4130140 | 1336389 | 187452 |\n",
       "| S3987Nr2-PBMC1_light | 3756935 | 3756935 | 3756935 | 1099571 | 170346 |\n",
       "| S3987Nr2-RAMOS_heavy | 2273707 | 2273707 | 2273707 | 511006  | 82177  |\n",
       "\n"
      ],
      "text/plain": [
       "  sample_name          raw     primer_pass pair_pass under2  atleast2\n",
       "1 S3987Nr1-PBMC1_heavy 293087  293087      293087    127350  14008   \n",
       "2 S3987Nr1-PBMC1_light 264459  264459      264459    108203  13384   \n",
       "3 S3987Nr1-RAMOS_heavy 160952  160952      160952    28      5       \n",
       "4 S3987Nr1-RAMOS_light 69991   69991       69991     25974   3791    \n",
       "5 S3987Nr2-PBMC1_heavy 4130140 4130140     4130140   1336389 187452  \n",
       "6 S3987Nr2-PBMC1_light 3756935 3756935     3756935   1099571 170346  \n",
       "7 S3987Nr2-RAMOS_heavy 2273707 2273707     2273707   511006  82177   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db6ff740-53a4-472b-93db-9dacb7d962ec",
   "metadata": {},
   "source": [
    "### Compare clone counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "957152e6-f19b-4ab8-a89c-1b65a0ead60a",
   "metadata": {},
   "outputs": [],
   "source": [
    "columnar_dataframe_builder <- function(values_vec, colnames) {\n",
    "    df = data.table::transpose(data.frame(values_vec))\n",
    "    colnames(df) = colnames\n",
    "    return(df)\n",
    "}\n",
    "\n",
    "aggregate_tsvs <- function(samples, path_builder, mode = \"headers\", nrows = 5, processor = NULL) {\n",
    "    \n",
    "    if (!(mode %in% c(\"headers\", \"processor\"))) {\n",
    "        stop(paste0(\"Mode not supported: \", mode))\n",
    "    }\n",
    "    \n",
    "    df_res = data.frame()\n",
    "    for (i in 1:length(samples)) {\n",
    "        \n",
    "        sample = samples[[i]]\n",
    "        filepath = path_builder(sample$name, sample$base_path)\n",
    "        \n",
    "        if (file.exists(filepath)) {\n",
    "            \n",
    "            if (mode == \"headers\") {\n",
    "                tsv_head = head(read_tsv(filepath), n = nrows)\n",
    "                df_annotated = cbind(sample$name, tsv_head)\n",
    "                df_res = plyr::rbind.fill(df_res, df_annotated)\n",
    "            } else if (mode == \"processor\") {\n",
    "                tsv = read_tsv(filepath)\n",
    "                df_processed = processor(tsv)\n",
    "                df_annotated = cbind(sample$name, df_processed)\n",
    "                df_res = plyr::rbind.fill(df_res, df_annotated)\n",
    "            }\n",
    "        \n",
    "        } else {\n",
    "            print(paste0(\"Filepath not found: \", filepath))\n",
    "        }\n",
    "    }\n",
    "    return(df_res)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "70b6c84f-9b77-473e-8b37-b231f64b9a36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 30 x 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>sample$name</th><th scope=col>clone_id</th><th scope=col>seq_count</th><th scope=col>copy_count</th><th scope=col>seq_freq</th><th scope=col>copy_freq</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 556</td><td>1628</td><td>24866</td><td>0.1248370524</td><td>1.944251e-01</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 374</td><td> 718</td><td>10748</td><td>0.0550571275</td><td>8.403769e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 900</td><td> 781</td><td> 8913</td><td>0.0598880454</td><td>6.968998e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 453</td><td> 533</td><td> 8079</td><td>0.0408710988</td><td>6.316901e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td>1316</td><td> 536</td><td> 7377</td><td>0.0411011426</td><td>5.768013e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td>1216</td><td> 333</td><td> 3103</td><td>0.0255348516</td><td>2.426209e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td>1259</td><td> 220</td><td> 1913</td><td>0.0168698719</td><td>1.495758e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 942</td><td> 206</td><td> 1856</td><td>0.0157963346</td><td>1.451190e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td>1144</td><td> 175</td><td> 1595</td><td>0.0134192163</td><td>1.247117e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 170</td><td> 175</td><td> 1557</td><td>0.0134192163</td><td>1.217405e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 148</td><td>2122</td><td>29546</td><td>0.1661316840</td><td>2.304375e-01</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 267</td><td> 764</td><td>10638</td><td>0.0598136695</td><td>8.296872e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 171</td><td> 667</td><td> 7790</td><td>0.0522195256</td><td>6.075637e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 195</td><td> 715</td><td> 7083</td><td>0.0559774524</td><td>5.524228e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 264</td><td> 777</td><td> 6767</td><td>0.0608314413</td><td>5.277771e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 301</td><td> 451</td><td> 5654</td><td>0.0353088546</td><td>4.409712e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 426</td><td> 444</td><td> 5461</td><td>0.0347608236</td><td>4.259186e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 436</td><td> 467</td><td> 5332</td><td>0.0365614969</td><td>4.158575e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 408</td><td> 356</td><td> 3828</td><td>0.0278712910</td><td>2.985564e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 218</td><td> 420</td><td> 3578</td><td>0.0328818602</td><td>2.790582e-02</td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_light</td><td>   1</td><td>2725</td><td>34447</td><td>0.9876766945</td><td>9.977119e-01</td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_light</td><td>  24</td><td>   5</td><td>   12</td><td>0.0018122508</td><td>3.475642e-04</td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_light</td><td>  27</td><td>   2</td><td>    7</td><td>0.0007249003</td><td>2.027458e-04</td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_light</td><td>   5</td><td>   1</td><td>    5</td><td>0.0003624502</td><td>1.448184e-04</td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_light</td><td>   8</td><td>   2</td><td>    4</td><td>0.0007249003</td><td>1.158547e-04</td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_light</td><td>  14</td><td>   1</td><td>    4</td><td>0.0003624502</td><td>1.158547e-04</td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_light</td><td>   7</td><td>   1</td><td>    3</td><td>0.0003624502</td><td>8.689104e-05</td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_light</td><td>   2</td><td>   1</td><td>    2</td><td>0.0003624502</td><td>5.792736e-05</td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_light</td><td>   3</td><td>   1</td><td>    2</td><td>0.0003624502</td><td>5.792736e-05</td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_light</td><td>   4</td><td>   1</td><td>    2</td><td>0.0003624502</td><td>5.792736e-05</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 30 x 6\n",
       "\\begin{tabular}{llllll}\n",
       " sample\\$name & clone\\_id & seq\\_count & copy\\_count & seq\\_freq & copy\\_freq\\\\\n",
       " <chr> & <int> & <int> & <int> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t S3987Nr1-PBMC1\\_heavy &  556 & 1628 & 24866 & 0.1248370524 & 1.944251e-01\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy &  374 &  718 & 10748 & 0.0550571275 & 8.403769e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy &  900 &  781 &  8913 & 0.0598880454 & 6.968998e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy &  453 &  533 &  8079 & 0.0408710988 & 6.316901e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy & 1316 &  536 &  7377 & 0.0411011426 & 5.768013e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy & 1216 &  333 &  3103 & 0.0255348516 & 2.426209e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy & 1259 &  220 &  1913 & 0.0168698719 & 1.495758e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy &  942 &  206 &  1856 & 0.0157963346 & 1.451190e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy & 1144 &  175 &  1595 & 0.0134192163 & 1.247117e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy &  170 &  175 &  1557 & 0.0134192163 & 1.217405e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  148 & 2122 & 29546 & 0.1661316840 & 2.304375e-01\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  267 &  764 & 10638 & 0.0598136695 & 8.296872e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  171 &  667 &  7790 & 0.0522195256 & 6.075637e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  195 &  715 &  7083 & 0.0559774524 & 5.524228e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  264 &  777 &  6767 & 0.0608314413 & 5.277771e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  301 &  451 &  5654 & 0.0353088546 & 4.409712e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  426 &  444 &  5461 & 0.0347608236 & 4.259186e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  436 &  467 &  5332 & 0.0365614969 & 4.158575e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  408 &  356 &  3828 & 0.0278712910 & 2.985564e-02\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  218 &  420 &  3578 & 0.0328818602 & 2.790582e-02\\\\\n",
       "\t S3987Nr1-RAMOS\\_light &    1 & 2725 & 34447 & 0.9876766945 & 9.977119e-01\\\\\n",
       "\t S3987Nr1-RAMOS\\_light &   24 &    5 &    12 & 0.0018122508 & 3.475642e-04\\\\\n",
       "\t S3987Nr1-RAMOS\\_light &   27 &    2 &     7 & 0.0007249003 & 2.027458e-04\\\\\n",
       "\t S3987Nr1-RAMOS\\_light &    5 &    1 &     5 & 0.0003624502 & 1.448184e-04\\\\\n",
       "\t S3987Nr1-RAMOS\\_light &    8 &    2 &     4 & 0.0007249003 & 1.158547e-04\\\\\n",
       "\t S3987Nr1-RAMOS\\_light &   14 &    1 &     4 & 0.0003624502 & 1.158547e-04\\\\\n",
       "\t S3987Nr1-RAMOS\\_light &    7 &    1 &     3 & 0.0003624502 & 8.689104e-05\\\\\n",
       "\t S3987Nr1-RAMOS\\_light &    2 &    1 &     2 & 0.0003624502 & 5.792736e-05\\\\\n",
       "\t S3987Nr1-RAMOS\\_light &    3 &    1 &     2 & 0.0003624502 & 5.792736e-05\\\\\n",
       "\t S3987Nr1-RAMOS\\_light &    4 &    1 &     2 & 0.0003624502 & 5.792736e-05\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 30 x 6\n",
       "\n",
       "| sample$name &lt;chr&gt; | clone_id &lt;int&gt; | seq_count &lt;int&gt; | copy_count &lt;int&gt; | seq_freq &lt;dbl&gt; | copy_freq &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| S3987Nr1-PBMC1_heavy |  556 | 1628 | 24866 | 0.1248370524 | 1.944251e-01 |\n",
       "| S3987Nr1-PBMC1_heavy |  374 |  718 | 10748 | 0.0550571275 | 8.403769e-02 |\n",
       "| S3987Nr1-PBMC1_heavy |  900 |  781 |  8913 | 0.0598880454 | 6.968998e-02 |\n",
       "| S3987Nr1-PBMC1_heavy |  453 |  533 |  8079 | 0.0408710988 | 6.316901e-02 |\n",
       "| S3987Nr1-PBMC1_heavy | 1316 |  536 |  7377 | 0.0411011426 | 5.768013e-02 |\n",
       "| S3987Nr1-PBMC1_heavy | 1216 |  333 |  3103 | 0.0255348516 | 2.426209e-02 |\n",
       "| S3987Nr1-PBMC1_heavy | 1259 |  220 |  1913 | 0.0168698719 | 1.495758e-02 |\n",
       "| S3987Nr1-PBMC1_heavy |  942 |  206 |  1856 | 0.0157963346 | 1.451190e-02 |\n",
       "| S3987Nr1-PBMC1_heavy | 1144 |  175 |  1595 | 0.0134192163 | 1.247117e-02 |\n",
       "| S3987Nr1-PBMC1_heavy |  170 |  175 |  1557 | 0.0134192163 | 1.217405e-02 |\n",
       "| S3987Nr1-PBMC1_light |  148 | 2122 | 29546 | 0.1661316840 | 2.304375e-01 |\n",
       "| S3987Nr1-PBMC1_light |  267 |  764 | 10638 | 0.0598136695 | 8.296872e-02 |\n",
       "| S3987Nr1-PBMC1_light |  171 |  667 |  7790 | 0.0522195256 | 6.075637e-02 |\n",
       "| S3987Nr1-PBMC1_light |  195 |  715 |  7083 | 0.0559774524 | 5.524228e-02 |\n",
       "| S3987Nr1-PBMC1_light |  264 |  777 |  6767 | 0.0608314413 | 5.277771e-02 |\n",
       "| S3987Nr1-PBMC1_light |  301 |  451 |  5654 | 0.0353088546 | 4.409712e-02 |\n",
       "| S3987Nr1-PBMC1_light |  426 |  444 |  5461 | 0.0347608236 | 4.259186e-02 |\n",
       "| S3987Nr1-PBMC1_light |  436 |  467 |  5332 | 0.0365614969 | 4.158575e-02 |\n",
       "| S3987Nr1-PBMC1_light |  408 |  356 |  3828 | 0.0278712910 | 2.985564e-02 |\n",
       "| S3987Nr1-PBMC1_light |  218 |  420 |  3578 | 0.0328818602 | 2.790582e-02 |\n",
       "| S3987Nr1-RAMOS_light |    1 | 2725 | 34447 | 0.9876766945 | 9.977119e-01 |\n",
       "| S3987Nr1-RAMOS_light |   24 |    5 |    12 | 0.0018122508 | 3.475642e-04 |\n",
       "| S3987Nr1-RAMOS_light |   27 |    2 |     7 | 0.0007249003 | 2.027458e-04 |\n",
       "| S3987Nr1-RAMOS_light |    5 |    1 |     5 | 0.0003624502 | 1.448184e-04 |\n",
       "| S3987Nr1-RAMOS_light |    8 |    2 |     4 | 0.0007249003 | 1.158547e-04 |\n",
       "| S3987Nr1-RAMOS_light |   14 |    1 |     4 | 0.0003624502 | 1.158547e-04 |\n",
       "| S3987Nr1-RAMOS_light |    7 |    1 |     3 | 0.0003624502 | 8.689104e-05 |\n",
       "| S3987Nr1-RAMOS_light |    2 |    1 |     2 | 0.0003624502 | 5.792736e-05 |\n",
       "| S3987Nr1-RAMOS_light |    3 |    1 |     2 | 0.0003624502 | 5.792736e-05 |\n",
       "| S3987Nr1-RAMOS_light |    4 |    1 |     2 | 0.0003624502 | 5.792736e-05 |\n",
       "\n"
      ],
      "text/plain": [
       "   sample$name          clone_id seq_count copy_count seq_freq     copy_freq   \n",
       "1  S3987Nr1-PBMC1_heavy  556     1628      24866      0.1248370524 1.944251e-01\n",
       "2  S3987Nr1-PBMC1_heavy  374      718      10748      0.0550571275 8.403769e-02\n",
       "3  S3987Nr1-PBMC1_heavy  900      781       8913      0.0598880454 6.968998e-02\n",
       "4  S3987Nr1-PBMC1_heavy  453      533       8079      0.0408710988 6.316901e-02\n",
       "5  S3987Nr1-PBMC1_heavy 1316      536       7377      0.0411011426 5.768013e-02\n",
       "6  S3987Nr1-PBMC1_heavy 1216      333       3103      0.0255348516 2.426209e-02\n",
       "7  S3987Nr1-PBMC1_heavy 1259      220       1913      0.0168698719 1.495758e-02\n",
       "8  S3987Nr1-PBMC1_heavy  942      206       1856      0.0157963346 1.451190e-02\n",
       "9  S3987Nr1-PBMC1_heavy 1144      175       1595      0.0134192163 1.247117e-02\n",
       "10 S3987Nr1-PBMC1_heavy  170      175       1557      0.0134192163 1.217405e-02\n",
       "11 S3987Nr1-PBMC1_light  148     2122      29546      0.1661316840 2.304375e-01\n",
       "12 S3987Nr1-PBMC1_light  267      764      10638      0.0598136695 8.296872e-02\n",
       "13 S3987Nr1-PBMC1_light  171      667       7790      0.0522195256 6.075637e-02\n",
       "14 S3987Nr1-PBMC1_light  195      715       7083      0.0559774524 5.524228e-02\n",
       "15 S3987Nr1-PBMC1_light  264      777       6767      0.0608314413 5.277771e-02\n",
       "16 S3987Nr1-PBMC1_light  301      451       5654      0.0353088546 4.409712e-02\n",
       "17 S3987Nr1-PBMC1_light  426      444       5461      0.0347608236 4.259186e-02\n",
       "18 S3987Nr1-PBMC1_light  436      467       5332      0.0365614969 4.158575e-02\n",
       "19 S3987Nr1-PBMC1_light  408      356       3828      0.0278712910 2.985564e-02\n",
       "20 S3987Nr1-PBMC1_light  218      420       3578      0.0328818602 2.790582e-02\n",
       "21 S3987Nr1-RAMOS_light    1     2725      34447      0.9876766945 9.977119e-01\n",
       "22 S3987Nr1-RAMOS_light   24        5         12      0.0018122508 3.475642e-04\n",
       "23 S3987Nr1-RAMOS_light   27        2          7      0.0007249003 2.027458e-04\n",
       "24 S3987Nr1-RAMOS_light    5        1          5      0.0003624502 1.448184e-04\n",
       "25 S3987Nr1-RAMOS_light    8        2          4      0.0007249003 1.158547e-04\n",
       "26 S3987Nr1-RAMOS_light   14        1          4      0.0003624502 1.158547e-04\n",
       "27 S3987Nr1-RAMOS_light    7        1          3      0.0003624502 8.689104e-05\n",
       "28 S3987Nr1-RAMOS_light    2        1          2      0.0003624502 5.792736e-05\n",
       "29 S3987Nr1-RAMOS_light    3        1          2      0.0003624502 5.792736e-05\n",
       "30 S3987Nr1-RAMOS_light    4        1          2      0.0003624502 5.792736e-05"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "clone_counts = aggregate_tsvs(\n",
    "    SAMPLES,\n",
    "    function(name, base_path) {get_path_clones(name, \"counts\", base_path)},\n",
    "    nrows = 10\n",
    ")\n",
    "clone_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5724c1c-8d0f-4973-86e2-042be109923f",
   "metadata": {},
   "source": [
    "### Abundancy and diversity comparisons among samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "5ce3600f-8ee2-4bf8-a3d3-6824b04d3895",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 21 x 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>sample$name</th><th scope=col>X1.order</th><th scope=col>coverages</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 1</td><td>0.9670290</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 2</td><td>0.9999302</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 3</td><td>0.9999999</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 4</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 5</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 6</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 7</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 8</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td> 9</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td>10</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 1</td><td>0.9870832</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 2</td><td>0.9999777</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 3</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 4</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 5</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 6</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 7</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 8</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 9</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td>10</td><td>1.0000000</td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_light</td><td> 1</td><td>0.9909393</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 21 x 3\n",
       "\\begin{tabular}{lll}\n",
       " sample\\$name & X1.order & coverages\\\\\n",
       " <chr> & <int> & <dbl>\\\\\n",
       "\\hline\n",
       "\t S3987Nr1-PBMC1\\_heavy &  1 & 0.9670290\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy &  2 & 0.9999302\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy &  3 & 0.9999999\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy &  4 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy &  5 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy &  6 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy &  7 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy &  8 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy &  9 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_heavy & 10 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  1 & 0.9870832\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  2 & 0.9999777\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  3 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  4 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  5 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  6 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  7 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  8 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  9 & 1.0000000\\\\\n",
       "\t S3987Nr1-PBMC1\\_light & 10 & 1.0000000\\\\\n",
       "\t S3987Nr1-RAMOS\\_light &  1 & 0.9909393\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 21 x 3\n",
       "\n",
       "| sample$name &lt;chr&gt; | X1.order &lt;int&gt; | coverages &lt;dbl&gt; |\n",
       "|---|---|---|\n",
       "| S3987Nr1-PBMC1_heavy |  1 | 0.9670290 |\n",
       "| S3987Nr1-PBMC1_heavy |  2 | 0.9999302 |\n",
       "| S3987Nr1-PBMC1_heavy |  3 | 0.9999999 |\n",
       "| S3987Nr1-PBMC1_heavy |  4 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_heavy |  5 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_heavy |  6 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_heavy |  7 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_heavy |  8 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_heavy |  9 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_heavy | 10 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_light |  1 | 0.9870832 |\n",
       "| S3987Nr1-PBMC1_light |  2 | 0.9999777 |\n",
       "| S3987Nr1-PBMC1_light |  3 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_light |  4 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_light |  5 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_light |  6 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_light |  7 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_light |  8 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_light |  9 | 1.0000000 |\n",
       "| S3987Nr1-PBMC1_light | 10 | 1.0000000 |\n",
       "| S3987Nr1-RAMOS_light |  1 | 0.9909393 |\n",
       "\n"
      ],
      "text/plain": [
       "   sample$name          X1.order coverages\n",
       "1  S3987Nr1-PBMC1_heavy  1       0.9670290\n",
       "2  S3987Nr1-PBMC1_heavy  2       0.9999302\n",
       "3  S3987Nr1-PBMC1_heavy  3       0.9999999\n",
       "4  S3987Nr1-PBMC1_heavy  4       1.0000000\n",
       "5  S3987Nr1-PBMC1_heavy  5       1.0000000\n",
       "6  S3987Nr1-PBMC1_heavy  6       1.0000000\n",
       "7  S3987Nr1-PBMC1_heavy  7       1.0000000\n",
       "8  S3987Nr1-PBMC1_heavy  8       1.0000000\n",
       "9  S3987Nr1-PBMC1_heavy  9       1.0000000\n",
       "10 S3987Nr1-PBMC1_heavy 10       1.0000000\n",
       "11 S3987Nr1-PBMC1_light  1       0.9870832\n",
       "12 S3987Nr1-PBMC1_light  2       0.9999777\n",
       "13 S3987Nr1-PBMC1_light  3       1.0000000\n",
       "14 S3987Nr1-PBMC1_light  4       1.0000000\n",
       "15 S3987Nr1-PBMC1_light  5       1.0000000\n",
       "16 S3987Nr1-PBMC1_light  6       1.0000000\n",
       "17 S3987Nr1-PBMC1_light  7       1.0000000\n",
       "18 S3987Nr1-PBMC1_light  8       1.0000000\n",
       "19 S3987Nr1-PBMC1_light  9       1.0000000\n",
       "20 S3987Nr1-PBMC1_light 10       1.0000000\n",
       "21 S3987Nr1-RAMOS_light  1       0.9909393"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "coverages_df = aggregate_tsvs(\n",
    "    SAMPLES, \n",
    "    function(sample_name, base_path) {get_path_coverages(sample_name, base_path)}, \n",
    "    nrows = 10\n",
    ")\n",
    "coverages_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "37e52fde-8f7b-44b1-9d82-c984ea3680ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "diversity <- read_tsv(\n",
    "    get_path_diversity(SAMPLES[[1]]$name, SAMPLES[[1]]$base_path)\n",
    ")\n",
    "\n",
    "interpret_diversity <- function(diversity_df) {\n",
    "\n",
    "    extract_div_by_q <- function(q) {diversity_df[diversity$q == q, ]$d}\n",
    "    \n",
    "    species_richness <- extract_div_by_q(0)\n",
    "    D_1 <- extract_div_by_q(1)\n",
    "    D_2 <- extract_div_by_q(2)\n",
    "    D_3 <- extract_div_by_q(3)\n",
    "    D_4 <- extract_div_by_q(4)\n",
    "    shannon_index <- log(D_1, 2)\n",
    "    gini_simpson_index <- 1 - 1/D_2\n",
    "    \n",
    "    res = columnar_dataframe_builder(\n",
    "        values_vec = c(species_richness, D_1, D_2, D_3, D_4, shannon_index, gini_simpson_index),\n",
    "        colnames = c(\"Species_richness_D_0\", \"D_1\", \"D_2\", \"D_3\", \"D_4\", \"Shannon_uncertainty\", \"Gini_simpson_prob\")\n",
    "    )\n",
    "    \n",
    "#     res = data.frame(\n",
    "#         c(\"Species_richness_D_0\", \"D_1\", \"D_2\", \"D_3\", \"D_4\", \"Shannon_uncertainty\", \"Gini_simpson_prob\"), \n",
    "#         c(species_richness, D_1, D_2, D_3, D_4, shannon_index, gini_simpson_index)\n",
    "#     )\n",
    "    return(res)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "39849325-07fd-4f18-8ba1-a9c18f37945a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3 x 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>sample$name</th><th scope=col>Species_richness_D_0</th><th scope=col>D_1</th><th scope=col>D_2</th><th scope=col>D_3</th><th scope=col>D_4</th><th scope=col>Shannon_uncertainty</th><th scope=col>Gini_simpson_prob</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>S3987Nr1-PBMC1_heavy</td><td>1254.24</td><td>213.581173</td><td>35.692533</td><td>20.000165</td><td>15.449315</td><td>7.7386407</td><td>0.97198294</td></tr>\n",
       "\t<tr><td>S3987Nr1-PBMC1_light</td><td> 540.08</td><td> 58.269088</td><td>20.190402</td><td>13.385689</td><td>10.718324</td><td>5.8646588</td><td>0.95047152</td></tr>\n",
       "\t<tr><td>S3987Nr1-RAMOS_light</td><td>  27.84</td><td>  1.109768</td><td> 1.024784</td><td> 1.018536</td><td> 1.016459</td><td>0.1502587</td><td>0.02418428</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 x 8\n",
       "\\begin{tabular}{llllllll}\n",
       " sample\\$name & Species\\_richness\\_D\\_0 & D\\_1 & D\\_2 & D\\_3 & D\\_4 & Shannon\\_uncertainty & Gini\\_simpson\\_prob\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t S3987Nr1-PBMC1\\_heavy & 1254.24 & 213.581173 & 35.692533 & 20.000165 & 15.449315 & 7.7386407 & 0.97198294\\\\\n",
       "\t S3987Nr1-PBMC1\\_light &  540.08 &  58.269088 & 20.190402 & 13.385689 & 10.718324 & 5.8646588 & 0.95047152\\\\\n",
       "\t S3987Nr1-RAMOS\\_light &   27.84 &   1.109768 &  1.024784 &  1.018536 &  1.016459 & 0.1502587 & 0.02418428\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 x 8\n",
       "\n",
       "| sample$name &lt;chr&gt; | Species_richness_D_0 &lt;dbl&gt; | D_1 &lt;dbl&gt; | D_2 &lt;dbl&gt; | D_3 &lt;dbl&gt; | D_4 &lt;dbl&gt; | Shannon_uncertainty &lt;dbl&gt; | Gini_simpson_prob &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| S3987Nr1-PBMC1_heavy | 1254.24 | 213.581173 | 35.692533 | 20.000165 | 15.449315 | 7.7386407 | 0.97198294 |\n",
       "| S3987Nr1-PBMC1_light |  540.08 |  58.269088 | 20.190402 | 13.385689 | 10.718324 | 5.8646588 | 0.95047152 |\n",
       "| S3987Nr1-RAMOS_light |   27.84 |   1.109768 |  1.024784 |  1.018536 |  1.016459 | 0.1502587 | 0.02418428 |\n",
       "\n"
      ],
      "text/plain": [
       "  sample$name          Species_richness_D_0 D_1        D_2       D_3      \n",
       "1 S3987Nr1-PBMC1_heavy 1254.24              213.581173 35.692533 20.000165\n",
       "2 S3987Nr1-PBMC1_light  540.08               58.269088 20.190402 13.385689\n",
       "3 S3987Nr1-RAMOS_light   27.84                1.109768  1.024784  1.018536\n",
       "  D_4       Shannon_uncertainty Gini_simpson_prob\n",
       "1 15.449315 7.7386407           0.97198294       \n",
       "2 10.718324 5.8646588           0.95047152       \n",
       "3  1.016459 0.1502587           0.02418428       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# head(diversity)\n",
    "# interpret_diversity(diversity)\n",
    "aggregate_tsvs(\n",
    "    SAMPLES,\n",
    "    get_path_diversity,\n",
    "    mode = \"processor\",\n",
    "    processor = interpret_diversity\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e476f782-5cf6-4362-9ce7-ec3e3b9fd300",
   "metadata": {},
   "source": [
    "### Join plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3689781a-97ec-4eab-83c1-648c07917100",
   "metadata": {},
   "outputs": [],
   "source": [
    "# magick::image_append(magick::image_scale(img, \"100\"), stack = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "071940bd-8042-4f34-9915-5cc74eeb0d90",
   "metadata": {},
   "outputs": [],
   "source": [
    "img1 = magick::image_read(get_path_pngs(SAMPLES[[1]]$name, \"abundancy\", SAMPLES[[1]]$base_path))\n",
    "img2 = magick::image_read(get_path_pngs(SAMPLES[[2]]$name, \"abundancy\", SAMPLES[[2]]$base_path))\n",
    "img3 = magick::image_read(get_path_pngs(SAMPLES[[3]]$name, \"abundancy\", SAMPLES[[3]]$base_path))\n",
    "comb_img = magick::image_append(c(img1, img2, img3), stack=TRUE)\n",
    "magick::image_write(comb_img, path = \"tiger.png\", format = \"png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "29a87f5a-92a6-4bc5-98a5-6db6a94bbdea",
   "metadata": {},
   "outputs": [],
   "source": [
    "img1 = magick::image_read(get_path_pngs(SAMPLES[[1]]$name, \"diversity\", SAMPLES[[1]]$base_path))\n",
    "img2 = magick::image_read(get_path_pngs(SAMPLES[[2]]$name, \"diversity\", SAMPLES[[2]]$base_path))\n",
    "img3 = magick::image_read(get_path_pngs(SAMPLES[[3]]$name, \"diversity\", SAMPLES[[3]]$base_path))\n",
    "# img4 = magick::image_read(get_path_pngs(SAMPLES[[4]]$name, \"clones\", SAMPLES[[4]]$base_path))\n",
    "\n",
    "comb_img = comb_img = magick::image_append(c(img1, img2, img3), stack=TRUE)\n",
    "magick::image_write(comb_img, path = \"tiger.png\", format = \"png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2dfdcf63-0846-478a-9004-e21a81b32d56",
   "metadata": {},
   "outputs": [],
   "source": [
    "img1 = magick::image_read(get_path_pngs(SAMPLES[[1]]$name, \"clones\", SAMPLES[[1]]$base_path))\n",
    "img2 = magick::image_read(get_path_pngs(SAMPLES[[2]]$name, \"clones\", SAMPLES[[2]]$base_path))\n",
    "img3 = magick::image_read(get_path_pngs(SAMPLES[[3]]$name, \"clones\", SAMPLES[[3]]$base_path))\n",
    "# img4 = magick::image_read(get_path_pngs(SAMPLES[[4]]$name, \"clones\", SAMPLES[[4]]$base_path))\n",
    "\n",
    "comb_img = comb_img = magick::image_append(c(img1, img2, img3), stack=TRUE)\n",
    "magick::image_write(comb_img, path = \"tiger.png\", format = \"png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55db8228-8756-45a0-9564-c774cfa4d54f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
